src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionServiceIntegrationTest.java

```diff
@@ -0,0 +1,56 @@
+package uk.gov.justice.hmpps.datacompliance.services.referral;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.jdbc.Sql;
+import uk.gov.justice.hmpps.datacompliance.IntegrationTest;
+import uk.gov.justice.hmpps.datacompliance.repository.jpa.repository.referral.ReferralResolutionRepository;
+import uk.gov.justice.hmpps.datacompliance.repository.jpa.repository.retention.RetentionCheckRepository;
+
+import javax.transaction.Transactional;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.concurrent.CompletableFuture.runAsync;
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.justice.hmpps.datacompliance.repository.jpa.model.referral.ReferralResolution.ResolutionStatus.PENDING;
+import static uk.gov.justice.hmpps.datacompliance.repository.jpa.model.referral.ReferralResolution.ResolutionStatus.RETAINED;
+import static uk.gov.justice.hmpps.datacompliance.repository.jpa.model.retention.RetentionCheck.Status.RETENTION_REQUIRED;
+
+class ReferralResolutionServiceIntegrationTest extends IntegrationTest {
+
+    @Autowired
+    private ReferralResolutionRepository referralResolutionRepository;
+
+    @Autowired
+    private RetentionCheckRepository retentionCheckRepository;
+
+    @Autowired
+    private ReferralResolutionService referralResolutionService;
+
+    @Test
+    @Sql("offender_deletion_batch.sql")
+    @Sql("offender_deletion_referral.sql")
+    @Sql("referral_resolution.sql")
+    @Sql("retention_check.sql")
+    void processUpdatedRetentionCheckResultsInRetention() {
+
+        assertThat(referralResolutionRepository.findById(1L).orElseThrow().getResolutionStatus()).isEqualTo(PENDING);
+
+        CompletableFuture.allOf(
+                runAsync(() -> updateAndProcessCheck(1L)),
+                runAsync(() -> updateAndProcessCheck(2L)))
+                .join();
+
+        assertThat(referralResolutionRepository.findById(1L).orElseThrow().getResolutionStatus()).isEqualTo(RETAINED);
```

Comment:
If the pessimistic lock isn't used, a race condition can mean that the status remains as PENDING.  I've proved this is the case by (locally) putting some carefully timed `Thread.sleep` s in so it always happens.  By introducing the lock though, this issue is resolved.

Reply:
I realise this is not easy... but as this test doesn't always fail, is there any way to make it?  For instance by using a SpyBean to reproduce the sleeps that you've proved it with locally? Just worried that someone could remove the lock and not realise (though at least you have fixed the bug - and not worth spending loads of time on this test).

Reply:
I have managed to ensure that the test will fail if the lock removed, good suggestion 👍